src/cargo/ops/registry/publish.rs

```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
-            bail!(
-                "`{}` cannot be published.\n\
-                 The registry `{}` is not listed in the `package.publish` value in Cargo.toml.",
-                pkg.name(),
-                reg_name
-            );
-        }
-    }
     // This is only used to confirm that we can create a token before we build the package.
     // This causes the credential provider to be called an extra time, but keeps the same order of errors.
-    let ver = pkg.version().to_string();
-    let operation = Operation::Read;
-
-    let reg_or_index = match opts.reg_or_index.clone() {
-        Some(RegistryOrIndex::Registry(_)) | None => {
-            publish_registry.map(RegistryOrIndex::Registry)
-        }
-        val => val,
-    };
     let source_ids = super::get_source_id(opts.gctx, reg_or_index.as_ref())?;
     let mut registry = super::registry(
         opts.gctx,
         &source_ids,
         opts.token.as_ref().map(Secret::as_deref),
         reg_or_index.as_ref(),
         true,
-        Some(operation).filter(|_| !opts.dry_run),
+        Some(Operation::Read).filter(|_| !opts.dry_run),
     )?;
-    verify_dependencies(pkg, &registry, source_ids.original)?;
 
-    // Prepare a tarball, with a non-suppressible warning if metadata
-    // is missing since this is being put online.
-    let tarball = ops::package_one(
+    // Validate all the packages before publishing any of them.
+    for (pkg, _) in &pkgs {
+        verify_dependencies(pkg, &registry, source_ids.original)?;
+    }
+
+    let specs = Packages::from_flags(
+        false,
+        vec![],
+        just_pkgs.iter().map(|p| p.name().to_string()).collect(),
+    )?;
+    let pkg_dep_graph = ops::cargo_package::package_with_dep_graph(
         ws,
-        pkg,
         &PackageOpts {
             gctx: opts.gctx,
             verify: opts.verify,
             list: false,
             check_metadata: true,
             allow_dirty: opts.allow_dirty,
-            to_package: Packages::Default,
+            to_package: specs,
             targets: opts.targets.clone(),
             jobs: opts.jobs.clone(),
             keep_going: opts.keep_going,
-            cli_features,
-            reg_or_index,
+            cli_features: opts.cli_features.clone(),
+            reg_or_index: reg_or_index.clone(),
         },
     )?;
 
-    if !opts.dry_run {
-        let hash = cargo_util::Sha256::new()
-            .update_file(tarball.file())?
-            .finish_hex();
-        let operation = Operation::Publish {
-            name: pkg.name().as_str(),
-            vers: &ver,
-            cksum: &hash,
-        };
-        registry.set_token(Some(auth::auth_token(
-            &opts.gctx,
-            &source_ids.original,
-            None,
-            operation,
-            vec![],
-            false,
-        )?));
-    }
+    let (mut order, mut ready) = PublishOrder::new(&pkg_dep_graph.graph);
+    let mut to_confirm: HashSet<_> = ready.iter().copied().collect();
```

Comment:
For readability, it feels weird that `to_confirm` has things that are `ready` but not published.